Initial Draft Outline for VGG19 (#57)
johko/computer-vision-course: course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md

# Very Deep Convolutional Networks for Large Scale Image Recognition (2014)

> 1. Introduction

Reviewer: ```suggestion
## Introduction
```

> The VGG architecture was developed in 2014 by Karen Simonyan and Andrew Zisserman from the Visual Geometry Group -and hence named VGG- at Oxford University. The model demonstrated significant improvements over the past models at that time- to be specific 2014 Imagenet challange also known as ILSVRC.

Reviewer: ```suggestion
The VGG architecture was developed in 2014 by Karen Simonyan and Andrew Zisserman from the Visual Geometry Group -and hence named VGG- at Oxford University. The model demonstrated significant improvements over the past models at that time, to be specific, 2014 ImageNet challenge also known as ILSVRC.
```

> The VGG architecture was developed in 2014 by Karen Simonyan and Andrew Zisserman from the Visual Geometry Group -and hence named VGG- at Oxford University. The model demonstrated significant improvements over the past models at that time- to be specific 2014 Imagenet challange also known as ILSVRC.

Reviewer 2: I think that we should write something about ImageNet Challenge (the task, the dataset size, etc)

Author: @michaelshekasta as I mentioned in https://github.com/johko/computer-vision-course/pull/57#issuecomment-1809489260 , I think it is better to have a separate page for Imagenet Challenge as all the CNN models will eventually refer to that informations.

> The VGG architecture was developed in 2014 by Karen Simonyan and Andrew Zisserman from the Visual Geometry Group at Oxford University. The authors detailed key advancements in convolutional neural networks for the ImageNet Large Scale Visual Recognition Challenge (ILSVRC). VGG demonstrated significant improvements with increased depth and repeated small (3X3) convolutional filters. It was a research project to systematically evaluate prior art configurations and identify key structural designs. 

Reviewer: Maybe have a less technical/academic introduction at first, i.e. general to specific, e.g. [this chapter on encoder models](https://huggingface.co/learn/nlp-course/en/chapter1/5?fw=pt#encoder-models). 
The beginning of the paragraph can be something like: VGG is one of the earliest pre-trained models with convolutions.

Reviewer 3: Like Merve mentions, I think we could condense this into just the juciest / most exciting bits! An alternative (personal preference) could be like:

"Following its development in 2014, VGG significantly improved [50%, some easy statistic etc] over [previous options] with increased depth and convolutional filters. What makes these new additions so important is...." -> (then maybe you could expand a bit, 1-3 precise ways it helps)

Author: @lunarflu @merveenoyan i tried to make it less technical in https://github.com/johko/computer-vision-course/pull/57/commits/b68fbac69251dabefbb6478295d4bce1b0ca74fb

> 2. VGG Network Architechture 

Reviewer: Headers should be looking like this
```suggestion
## VGG Network Architechture 
```

> - Input is 224X224 image. 

Reviewer: ```suggestion
- Inputs are 224X224 images. 
```

> - 3X3 convolution kernels with 2X2 max pooling layers.

Reviewer: ```suggestion
- Convolution kernel shape is (3,3) and max pooling window shape is (2,2).
```

- Number of channels for each convolutional layer 64 -> 128 -> 256 -> 512 -> 512. 
- VGG16 has 16 hidden layers (13 convolutional layers and 3 fully connected layers).
- VGG19 has 19 hidden layers (16 convolutional layers and 3 fully connected layers).

Below is the network structure of VGG16, The activation and dropout layers were omitted for brevity. MaxPooling layers were included to display changes in output tensor size at each layer. 

> |Layer Type |Filter Size|Output Size|No. Params|

Reviewer: I think this is too in-depth.

Author: @merveenoyan sorry, i am confused, should i update this section?

Reviewer: Maybe you could remove the architecture table itself, that's what I meant

|-----------|------------|----------|----------|
|Input Image|0|3X224X224|0|
|Conv2D|3X3|64X224X224|1,792|
|Conv2D|3X3|64X224X224|39,628|
|MaxPool2D|2X2|64X112X112|0|
|Conv2D|3X3|128X112X112|73,856|
|Conv2D|3X3|128X112X112|147,584|
|MaxPool2D|2X2|128X56X56|0|
|Conv2D|3X3|256X56X56|295,168|
|Conv2D|3X3|256X56X56|590,080|
|Conv2D|3X3|256X56X56|590,080|
|MaxPool2D|2X2|256X28X28|0|
|Conv2D|3X3|512X28X28|1,180,160|
|Conv2D|3X3|512X28X28|2,359,808|
|Conv2D|3X3|512X28X28|2,359,808|
|MaxPool2D|2X2|512X14X14|0|
|Conv2D|3X3|512X14X14|2,359,808|
|Conv2D|3X3|512X14X14|2,359,808|
|Conv2D|3X3|512X14X14|2,359,808|
|MaxPool2D|2X2|512X7X7|0|
|Linear|N/A|4096|102,764,544|
|Linear|N/A|4096|16,781,312|

> |Linear|N/A|1000|4,096,000|

Reviewer 2: Do you think that we should add a graph or illustration?
e.g. https://neurohive.io/wp-content/uploads/2018/11/vgg16-1-e1542731207177.png

Author: @michaelshekasta we plan to add illustration in a new PR. can you resolve it for now?

> 3. Key Comparisons 

Reviewer: ```suggestion
## Key Comparisons 
```

> 3. Key Comparisons 

Reviewer: Very nice section :)

> - VGG (16 or 19 layers) is relatively deeper than other SOTA networks at the time. AlexNet, winning model for ILSVRC 2012 only has 8 layers.    

Reviewer 3: `AlexNet, the winning model`

Author: updated in https://github.com/johko/computer-vision-course/pull/57/commits/b68fbac69251dabefbb6478295d4bce1b0ca74fb

> - VGG (16 or 19 layers) is relatively deeper than other SOTA networks at the time. AlexNet, winning model for ILSVRC 2012 only has 8 layers.    

Reviewer 3: `was relatively deeper`

Author: updated in https://github.com/johko/computer-vision-course/pull/57/commits/b68fbac69251dabefbb6478295d4bce1b0ca74fb

- Multiple small (3X3) receptive field filters with ReLU activation instead of one large (7X7 or 11X11) filter lead to better learning of complex features. Smaller filters also mean fewer parameters per layer, with additional nonlinearity introduced in between.
- Multiscale training and inference. Each image was trained in multiple rounds with varying scales to ensure similar characteristics were captured at different sizes.  
- Consistency and simplicity of the VGG network make it easier to scale or modify for future improvements.

> 4. Pytorch Example

Reviewer: ```suggestion
## PyTorch Example
```
